kubeadm: restore and generalize the TestMigrateOldConfig test

neolit123 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind cleanup

#### What this PR does / why we need it:
```
In the past, the test required two APIs to be available to test for migration, which made us comment-out the test once v1beta2 was removed. Currently it just accepts a single (old) API to migrate. Keep it simple and use a variable "gv" on top of the function body to easily swap the version to be tested once an old API is deleted. e.g. currently v1beta3 is the "old" API, v1beta4 is the "new" one.

Ultimately, this test only makes sure that the expected kinds are available post migration.
```

```
Place the validate/migrate tests in common_tests.go on the backend side in
common_tests.go. A test for migrate TestMigrateOldConfig
was already present there. Apply slightly better coverage to it
and rename desc -> name.

Fix typo in argument oldConfig -> config in ValidateConfig().
```

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
xref https://github.com/kubernetes/kubeadm/issues/2890

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

neolit123 committed:
kubeadm: restore and generalize the TestMigrateOldConfig test

The test required two APIs to be available to test for migration.
Keep it simple and use a variable "gv" on top of the function body
to easily swap the version to be tested once an old API is deleted.
e.g. currently v1beta3 is the "old" API, v1beta4 is the "new" one.

Ultimately, this test only makes sure that the expected kinds are
available post migration.

neolit123 commented:
/triage accepted
/priority backlog

SataQiu commented on cmd/kubeadm/app/util/config/common_test.go:
```diff
@@ -213,209 +216,221 @@ func TestVerifyAPIServerBindAddress(t *testing.T) {
-// }
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			const allowExperimental = true
```
Do we also need to test the scenarios where `allowExperimental` is false?
How about adding `allowExperimental bool` as an input variable of the test cases?

neolit123 replied:
added `allowExperimental bool`

this was actually wrong:
> Migration of experimental APIs is already tested in the config_test.go file in the "cmd" package and not in this test.

experimental API migration is now tested only in this test function in terms of output kinds.
TestValidateSupportedVersion also tests if the input API is allowed as experimental in general.

our tests are a bit sparce between migrate and validate, but i think we should try to keep it simple and not update many files / tests cases when e.g. v1beta4 is no longer experimental.

we could remove the tests from cmd/config_test.go and do all the migrate / validate tests in common_test.go on the backend layer.

neolit123 replied:
i can send a PR follow up.
it can remove the cmd/config_tests.go tests and add test for ValidateConfig in common_test.go. also see if more cases are needed for MigrateOldConfig and add them in common_test.go

neolit123 replied:
> i can send a PR follow up.

actually, seems easier better as a separate commit here.
will update this PR a bit later.

neolit123 replied:
added second commit

neolit123 committed:
kubeadm: move migrate / validate tests out of config_tests.go

Place the tests in common_tests.go on the backend side in
common_tests.go. A test for migrate TestMigrateOldConfig
was already present there. Apply slightly better coverage to it
and rename desc -> name.

Fix typo in argument oldConfig -> config in ValidateConfig().

pacoxu commented on cmd/kubeadm/app/util/config/common_test.go:
```diff
@@ -439,6 +447,94 @@ func TestMigrateOldConfig(t *testing.T) {
+			cfg: dedent.Dedent(fmt.Sprintf(`
+			apiVersion: %s
+			kind: InitConfiguration
+			`, gvExperimental)),
```
A test case that is invalid due to `Experimental` not allowed, may be needed.

neolit123 replied:
good catch, added experimental=false, expectedError=true test cases for both validate and migrate.

SataQiu replied:
LGTM

pacoxu reviewed (commented):
Overall LGTM

SataQiu commented:
/lgtm
/approve
